File: golang/earthquakes.go
// To parse and unparse this JSON data, add this code to your project and do:
//
//    earthquakes, err := UnmarshalEarthquakes(bytes)
//    bytes, err = earthquakes.Marshal()

package main

import "encoding/json"

func UnmarshalEarthquakes(data []byte) (Earthquakes, error) {
	var r Earthquakes
	err := json.Unmarshal(data, &r)
	return r, err
}

func (r *Earthquakes) Marshal() ([]byte, error) {
	return json.Marshal(r)
}

type Earthquakes struct {
	Type     string    `json:"type"`    
	Metadata Metadata  `json:"metadata"`
	Features []Feature `json:"features"`
	Bbox     []float64 `json:"bbox"`    
}

type Feature struct {
	Type       string     `json:"type"`      
	Properties Properties `json:"properties"`
	Geometry   Geometry   `json:"geometry"`  
	ID         string     `json:"id"`        
}

type Geometry struct {
	Type        string    `json:"type"`       
	Coordinates []float64 `json:"coordinates"`
}

type Properties struct {
	Mag     float64     `json:"mag"`    
	Place   string      `json:"place"`  
	Time    int64       `json:"time"`   
	Updated int64       `json:"updated"`
	Tz      int64       `json:"tz"`     
	URL     string      `json:"url"`    
	Detail  string      `json:"detail"` 
	Felt    interface{} `json:"felt"`   
	Cdi     interface{} `json:"cdi"`    
	MMI     interface{} `json:"mmi"`    
	Alert   interface{} `json:"alert"`  
	Status  string      `json:"status"` 
	Tsunami int64       `json:"tsunami"`
	Sig     int64       `json:"sig"`    
	Net     string      `json:"net"`    
	Code    string      `json:"code"`   
	IDS     string      `json:"ids"`    
	Sources string      `json:"sources"`
	Types   string      `json:"types"`  
	Nst     int64       `json:"nst"`    
	Dmin    float64     `json:"dmin"`   
	RMS     float64     `json:"rms"`    
	Gap     int64       `json:"gap"`    
	MagType string      `json:"magType"`
	Type    string      `json:"type"`   
	Title   string      `json:"title"`  
}

type Metadata struct {
	Generated int64  `json:"generated"`
	URL       string `json:"url"`      
	Title     string `json:"title"`    
	Status    int64  `json:"status"`   
	API       string `json:"api"`      
	Count     int64  `json:"count"`    
}

File: javascript/earthquakes.js
// To parse this data:
//
//   const Convert = require("./file");
//
//   const earthquakes = Convert.toEarthquakes(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
function toEarthquakes(json) {
    return cast(JSON.parse(json), o("Earthquakes"));
}

function earthquakesToJson(value) {
    return JSON.stringify(value, null, 2);
}

function cast(obj, typ) {
    if (!isValid(typ, obj)) {
        throw `Invalid value`;
    }
    return obj;
}

function isValid(typ, val) {
    if (typ === undefined) return true;
    if (typ === null) return val === null || val === undefined;
    return typ.isUnion  ? isValidUnion(typ.typs, val)
            : typ.isArray  ? isValidArray(typ.typ, val)
            : typ.isMap    ? isValidMap(typ.typ, val)
            : typ.isEnum   ? isValidEnum(typ.name, val)
            : typ.isObject ? isValidObject(typ.cls, val)
            :                isValidPrimitive(typ, val);
}

function isValidPrimitive(typ, val) {
    return typeof typ === typeof val;
}

function isValidUnion(typs, val) {
    // val must validate against one typ in typs
    return typs.find(typ => isValid(typ, val)) !== undefined;
}

function isValidEnum(enumName, val) {
    const cases = typeMap[enumName];
    return cases.indexOf(val) !== -1;
}

function isValidArray(typ, val) {
    // val must be an array with no invalid elements
    return Array.isArray(val) && val.every(element => {
        return isValid(typ, element);
    });
}

function isValidMap(typ, val) {
    if (val === null || typeof val !== "object" || Array.isArray(val)) return false;
    // all values in the map must be typ
    return Object.keys(val).every(prop => {
        if (!Object.prototype.hasOwnProperty.call(val, prop)) return true;
        return isValid(typ, val[prop]);
    });
}

function isValidObject(className, val) {
    if (val === null || typeof val !== "object" || Array.isArray(val)) return false;
    let typeRep = typeMap[className];
    return Object.keys(typeRep).every(prop => {
        if (!Object.prototype.hasOwnProperty.call(typeRep, prop)) return true;
        return isValid(typeRep[prop], val[prop]);
    });
}

function a(typ) {
    return { typ, isArray: true };
}

function e(name) {
    return { name, isEnum: true };
}

function u(...typs) {
    return { typs, isUnion: true };
}

function m(typ) {
    return { typ, isMap: true };
}

function o(className) {
    return { cls: className, isObject: true };
}

const typeMap = {
    "Earthquakes": {
        type: "",
        metadata: o("Metadata"),
        features: a(o("Feature")),
        bbox: a(3.14),
    },
    "Feature": {
        type: "",
        properties: o("Properties"),
        geometry: o("Geometry"),
        id: "",
    },
    "Geometry": {
        type: "",
        coordinates: a(3.14),
    },
    "Properties": {
        mag: 3.14,
        place: "",
        time: 0,
        updated: 0,
        tz: 0,
        url: "",
        detail: "",
        felt: null,
        cdi: null,
        mmi: null,
        alert: null,
        status: "",
        tsunami: 0,
        sig: 0,
        net: "",
        code: "",
        ids: "",
        sources: "",
        types: "",
        nst: 0,
        dmin: 3.14,
        rms: 3.14,
        gap: 0,
        magType: "",
        type: "",
        title: "",
    },
    "Metadata": {
        generated: 0,
        url: "",
        title: "",
        status: 0,
        api: "",
        count: 0,
    },
};

module.exports = {
    "earthquakesToJson": earthquakesToJson,
    "toEarthquakes": toEarthquakes,
};
